TheMovieDb/Home/Contoller/HomeViewController.swift

```diff
@@ -7,21 +7,222 @@
     
     // MARK: - Life Cycle
+    
+    init() {
+        super.init(collectionViewLayout: UICollectionViewController.configureCollectionViewLayout())
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
     override func viewDidLoad() {
         super.viewDidLoad()
         configureUI()
-        // Do any additional setup after loading the view.
+        configureUICollection()
+        fetchDataAPI()
+
+    }
+    // MARK: - API
+    private func reloadCollectionView() {
+        DispatchQueue.main.async {
+            self.collectionView.reloadData()
+        }
+    }
+    
+    private func fetchDataAPI() {
+        fetchPopular()
+        fetchTopRated()
+        fetchNowPlaying()
+        fetchUpcoming()
+        fetchTrending()
+        group.notify(queue: .main) {
+            self.reloadCollectionView()
+        }
     }
     
+    private func fetchPopular() {
+        group.enter()
+        MovieAPI.shared.getPopular(completion: {(response: Result<Movies, Error>) in
+            switch response {
+            case .failure(let error):
+                debugPrint(error)
+            case .success(let res):
+                self.popularMovies = res.movies
+            }
+            self.group.leave()
+        })
+    }
+    
+    private func fetchTopRated() {
+        group.enter()
+        MovieAPI.shared.getTopRated(completion: {(response: Result<Movies, Error>) in
+            switch response {
+            case .failure(let error):
+                debugPrint(error)
+            case .success(let res):
+                self.topRatedMovies = res.movies
+            }
+            self.group.leave()
+        })
+    }
+    
+    private func fetchNowPlaying() {
+        group.enter()
+        MovieAPI.shared.getNowPlaying(completion: {(response: Result<Movies, Error>) in
+            switch response {
+            case .failure(let error):
+                debugPrint(error)
+            case .success(let res):
+                self.playingNowMovies = res.movies
+            }
+            self.group.leave()
+        })
+    }
+    
+    private func fetchUpcoming() {
+        group.enter()
+        MovieAPI.shared.getUpcoming(completion: {(response: Result<Movies, Error>) in
+            switch response {
+            case .failure(let error):
+                debugPrint(error)
+            case .success(let res):
+                self.upComingMovies =  res.movies
+            }
+            self.group.leave()
+        })
+    }
+    
+    private func fetchTrending() {
+        group.enter()
+         MovieAPI.shared.getTrending(completion: {(response: Result<Movies, Error>) in
+             switch response {
+             case .failure(let error):
+                 debugPrint(error)
+             case .success(let res):
+                 self.trendingMovies = res.movies
+             }
+             self.group.leave()
+         })
+    }
```

Comment:
all this func are repeating code, so it doesn't conform with DRY.
in fact the only thing that change is "self.[Movies]"

Reply:
fixed using generic method